File: BinarySort.py
def tachSo0VaSo1(arr):
    viTriDau = 0  # Chỉ số bắt đầu
    viTriCuoi = len(arr) - 1  # Chỉ số kết thúc

    # Duyệt danh sách để sắp xếp các số 0 về đầu và số 1 về cuối
    while viTriDau < viTriCuoi:
        if arr[viTriDau] == 1:  # Nếu phần tử ở đầu là 1
            if arr[viTriCuoi] != 1:  # Nếu phần tử ở cuối không phải là 1
                # Hoán đổi giá trị
                arr[viTriDau], arr[viTriCuoi] = arr[viTriCuoi], arr[viTriDau]
                viTriDau += 1  # Tăng chỉ số bắt đầu
                viTriCuoi -= 1  # Giảm chỉ số kết thúc
            else:
                viTriCuoi -= 1  # Giảm chỉ số kết thúc khi phần tử cuối là 1
        else:
            viTriDau += 1  # Tăng chỉ số bắt đầu khi phần tử đầu là 0
try:
    arr = list(map(int, input("Nhập danh sách các số 0 và 1, cách nhau bằng dấu cách: ").split()))
    if all(x in (0, 1) for x in arr):  # Kiểm tra danh sách chỉ chứa số 0 và 1
        tachSo0VaSo1(arr)
        print("Danh sách sau khi tách là:", arr)
    else:
        print("Danh sách chỉ được chứa các số 0 và 1.")
except ValueError:
    print("Vui lòng nhập danh sách hợp lệ.")

File: BinarySort.c
#include <stdio.h>

void tachSo0VaSo1(int mang[], int kichThuoc) {
    int viTriDau = 0;      // Chỉ số bắt đầu
    int viTriCuoi = kichThuoc - 1; // Chỉ số kết thúc

    while (viTriDau < viTriCuoi) {
        if (mang[viTriDau] == 1) {
            if (mang[viTriCuoi] != 1) {
                // Hoán đổi giá trị
                int tam = mang[viTriDau];
                mang[viTriDau] = mang[viTriCuoi];
                mang[viTriCuoi] = tam;
                viTriDau++;
                viTriCuoi--;
            } else {
                viTriCuoi--;
            }
        } else {
            viTriDau++;
        }
    }
}

int main() {
    int kichThuoc;
    printf("Nhập kích thước của mảng: ");
    scanf("%d", &kichThuoc);

    if (kichThuoc <= 0) {
        printf("Kích thước mảng phải lớn hơn 0.\n");
        return 1;
    }
    int mang[kichThuoc];
    printf("Nhập các phần tử của mảng (chỉ chứa 0 và 1): ");
    for (int i = 0; i < kichThuoc; i++) {
        scanf("%d", &mang[i]);
        // Kiểm tra giá trị hợp lệ
        if (mang[i] != 0 && mang[i] != 1) {
            printf("Mảng chỉ được chứa các số 0 và 1.\n");
            return 1;
        }
    }
    tachSo0VaSo1(mang, kichThuoc);
    printf("Mảng sau khi tách: ");
    for (int i = 0; i < kichThuoc; i++) {
        printf("%d ", mang[i]);
    }
    printf("\n");

    return 0;
}
